File: BlotoutAnalytics/Storage/BOAFileStorage.m
//
//  BOAFileStorage.m
//  BlotoutAnalytics
//
//  Copyright © 2021 Blotout. All rights reserved.
//

#import "BOAFileStorage.h"
#import "BOFLogs.h"
#import "BOFFileSystemManager.h"

@interface BOAFileStorage ()

@property (nonatomic, strong, nonnull) NSURL *folderURL;

@end

@implementation BOAFileStorage
- (instancetype)init {
  return [self initWithFolder:[NSURL fileURLWithPath:[BOFFileSystemManager getBOSDKRootDirectory]] crypto:nil];
}

- (instancetype)initWithFolder:(NSURL *)folderURL crypto:(id<BOACrypto>)crypto {
  if (self = [super init]) {
    _folderURL = folderURL;
    _crypto = crypto;
    [self createDirectoryAtURLIfNeeded:folderURL];
    return self;
  }
  return nil;
}

- (void)removeKey:(NSString *)key {
  NSURL *url = [self urlForKey:key];
  NSError *error = nil;
  if (![[NSFileManager defaultManager] removeItemAtURL:url error:&error]) {
    BOFLogDebug(@"Unable to remove key %@ - error removing file at path %@", key, url);
  }
}

- (void)resetAll {
  NSError *error = nil;
  if (![[NSFileManager defaultManager] removeItemAtURL:self.folderURL error:&error]) {
    BOFLogDebug(@"ERROR: Unable to reset file storage. Path cannot be removed - %@", self.folderURL.path);
  }
  
  [self createDirectoryAtURLIfNeeded:self.folderURL];
}

- (void)setData:(NSData *)data forKey:(NSString *)key {
  NSURL *url = [self urlForKey:key];
  if (self.crypto) {
    NSData *encryptedData = [self.crypto encrypt:data];
    [encryptedData writeToURL:url atomically:YES];
  } else {
    [data writeToURL:url atomically:YES];
  }
  
  NSError *error = nil;
  if (![url setResourceValue:@YES
                      forKey:NSURLIsExcludedFromBackupKey
                       error:&error]) {
    BOFLogDebug(@"Error excluding %@ from backup %@", [url lastPathComponent], error);
  }
}

- (NSData *)dataForKey:(NSString *)key {
  NSURL *url = [self urlForKey:key];
  NSData *data = [NSData dataWithContentsOfURL:url];
  if (!data) {
    BOFLogDebug(@"WARNING: No data file for key %@", key);
    return nil;
  }
  
  if (self.crypto) {
    return [self.crypto decrypt:data];
  }
  
  return data;
}

- (NSDictionary *)dictionaryForKey:(NSString *)key {
  return [self plistForKey:key];
}

- (void)setDictionary:(NSDictionary *)dictionary forKey:(NSString *)key {
  [self setPlist:dictionary forKey:key];
}

- (NSArray *)arrayForKey:(NSString *)key {
  return [self plistForKey:key];
}

- (void)setArray:(NSArray *)array forKey:(NSString *)key {
  [self setPlist:array forKey:key];
}

- (NSString *)stringForKey:(NSString *)key {
  return [self plistForKey:key];
}

- (void)setString:(NSString *)string forKey:(NSString *)key {
  [self setPlist:string forKey:key];
}

- (NSURL *)urlForKey:(NSString *)key {
  return [self.folderURL URLByAppendingPathComponent:key];
}

#pragma mark - Helpers

- (id _Nullable)plistForKey:(NSString *)key {
  NSData *data = [self dataForKey:key];
  return data ? [self plistFromData:data] : nil;
}

- (void)setPlist:(id _Nonnull)plist forKey:(NSString *)key {
  NSData *data = [self dataFromPlist:plist];
  if (data) {
    [self setData:data forKey:key];
  }
}

- (NSData *_Nullable)dataFromPlist:(nonnull id)plist {
    @try{
        NSError *error = nil;
        NSData *data = [NSPropertyListSerialization dataWithPropertyList:plist
                                                                  format:NSPropertyListXMLFormat_v1_0
                                                                 options:0
                                                                   error:&error];
        if (error) {
            BOFLogDebug(@"Unable to serialize data from plist object", error, plist);
        }
        
        return data;
    }@catch(NSException *exception) {
        return nil;
    }
    
}

- (id _Nullable)plistFromData:(NSData *_Nonnull)data {
  NSError *error = nil;
  id plist = [NSPropertyListSerialization propertyListWithData:data
                                                       options:0
                                                        format:nil
                                                         error:&error];
  if (error) {
    BOFLogDebug(@"Unable to parse plist from data %@", error);
  }
  
  return plist;
}

- (void)createDirectoryAtURLIfNeeded:(NSURL *)url {
  if ([[NSFileManager defaultManager] fileExistsAtPath:url.path
                                            isDirectory:NULL]) {
    return;
  }
  
  NSError *error = nil;
  if (![[NSFileManager defaultManager] createDirectoryAtPath:url.path
                                 withIntermediateDirectories:YES
                                                  attributes:nil
                                                       error:&error]) {
    BOFLogDebug(@"error: %@", error.localizedDescription);
  }
}

@end

File: BlotoutAnalytics/Storage/BOAFileStorage.swift
//
//  BOAFileStorage.swift
//  BlotoutAnalytics
//
//  Created by Poonam Tiwari on 19/04/22.
//

import Foundation

class BOAFileStorage:NSObject, BOAStorage {
    

    var folderURL: URL!

    convenience override init() {
        self.init(folder: URL(fileURLWithPath: BOFFileSystemManager.getBOSDKRootDirectory() ?? ""))
    }

    init(folder folderURL: URL) {
        super.init()
        self.folderURL = folderURL
        createDirectoryAtURLIfNeeded(url: folderURL)
    }
    
    func removeKey(_ key: String) {
        let url = self.urlForKey(key)
        do {
            if let removeURL = url {
                try FileManager.default.removeItem(at: removeURL)
            }
        } catch {
            BOFLogDebug(frmt: "Unable to remove key %@ - error removing file at path %@", args: key , url?.absoluteString as! CVarArg)
        }
    }

    func resetAll() {
        do {
            try FileManager.default.removeItem(at: folderURL)
        } catch {
            BOFLogDebug(frmt: "ERROR: Unable to reset file storage. Path cannot be removed - %@", args: folderURL.path)
        }

        createDirectoryAtURLIfNeeded(url: folderURL)
    }


    func setData(_ data: Data, forKey key: String) {
        let url = self.urlForKey( key)
        if let url = url {
            NSData(data: data).write(to: url, atomically: true)
        }
        do {
            try (url as NSURL?)?.setResourceValue(NSNumber(value: true), forKey: .isExcludedFromBackupKey)
        } catch {
            BOFLogDebug(frmt: "Error excluding %@ from backup %@", args: url?.lastPathComponent as! CVarArg, error.localizedDescription)
        }
    }
    
    func dataForKey(_ key: String) -> Data? {
        let url = self.urlForKey(key)
        var data: Data? = nil
        do{
            if let url = url {
                data = try Data(contentsOf: url) }
        }
        catch
        {
            BOFLogDebug(frmt: "WARNING: No data file for key %@", args: key)
            
        }
        if data == nil {
            BOFLogDebug(frmt: "WARNING: No data file for key %@", args: key)
            return nil
        }
        return data
    }
    
    func dictionaryForKey(_ key: String) -> [String : Any]? {
        return plistForKey(key) as? [String : Any]
    }
    
    
    func setDictionary(_ dictionary: [String : Any], forKey key: String) {
        setPlist(dictionary, forKey: key)
    }

    func arrayForKey(_ key: String) -> [EventModel]? {
        return plistForKey(key) as? [EventModel]
    }
    
    func setArray(_ array: [Any]?, forKey key: String) {
        setPlist(array, forKey: key)
    }

    func stringForKey(_ key: String) -> String? {
        return plistForKey(key) as? String
    }

    func setString(_ string: String?, forKey key: String) {
        setPlist(string, forKey: key)
    }
    
    func urlForKey(_ key: String?) -> URL? {
        return folderURL.appendingPathComponent(key ?? "")
    }

    // MARK: - Helpers

    
    func plistForKey(_ key: String) -> Any? {
        let data = self.dataForKey(key)
        return data != nil ? plistFromData(_:data!) : nil
    }
    
    func setPlist(_ plist: Any, forKey key: String) {
        let data = self.dataFromPlist(plist: plist)
        if let data = data {
            setData(data, forKey: key)
        }
    }
    
    func dataFromPlist(plist: Any) -> Data? {
        var data: Data? = nil
        do {
            data = try PropertyListSerialization.data(
                fromPropertyList: plist,
                format: .xml,
                options: 0)
            
            return data
            
        } catch {
            BOFLogDebug(frmt: "Unable to serialize data from plist object", args: error.localizedDescription as! CVarArg, plist as! CVarArg)
            return nil
        }
        //TODO: check condition
    }

    func plistFromData(_ data: Data) -> Any? {
        var plist: Any? = nil
        do {
            plist = try PropertyListSerialization.propertyList(
                from: data,
                options: [],
                format: nil)
        } catch {
            BOFLogDebug(frmt: "Unable to parse plist from data %@", args: error.localizedDescription)
        }
        return plist
    }
    
    func createDirectoryAtURLIfNeeded( url: URL) {
        if FileManager.default.fileExists(
            atPath: url.path ,
            isDirectory: nil) {
            return
        }

        do {
            try FileManager.default.createDirectory(
                atPath: url.path,
                withIntermediateDirectories: true,
                attributes: nil)
        } catch {
            BOFLogDebug(frmt: "error: %@", args: error.localizedDescription)
        }
    }
}

